src/socket.c

```diff
@@ -0,0 +1,263 @@
+/*
+ *------------------------------------------------------------------
+ * Copyright (c) 2017 Cisco and/or its affiliates.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *------------------------------------------------------------------
+ */
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <memif_private.h>
+#include <memif.h>
+
+#include <fifo.h>
+
+#define memif_min(a,b) a < b ? a : b
+
+/* fifo is used as message queue, fifo will probably remain as it is
```

Comment:
for message queue please use linked list instead of fifo. i.e.:
typedef struct memif_msg_queue_elt_ {
  memif_msg_t msg;
  int fd;
  struct memif_msg_queue_elt_ *next;
} memif_msg_queue_elt_t;